bsp4s/src/main/scala/ch/epfl/scala/bsp/Bsp.scala

```diff
@@ -627,160 +932,300 @@ object ScalaPlatform {
     target: BuildTargetIdentifier,
     // Fully qualified names of test classes
     classes: List[String]
 )
 
-@JsonCodec final case class ScalaTestClassesResult(
+object ScalaTestClassesItem {
+  implicit val codec: JsonValueCodec[ScalaTestClassesItem] =
+    JsonCodecMaker.make(CodecMakerConfig.withTransientEmpty(false))
+}
+
+final case class ScalaTestClassesResult(
     items: List[ScalaTestClassesItem]
 )
 
+object ScalaTestClassesResult {
+  implicit val codec: JsonValueCodec[ScalaTestClassesResult] =
+    JsonCodecMaker.make(CodecMakerConfig.withTransientEmpty(false))
+}
+
 // Request: 'buildTarget/scalaMainClasses', C -> S
-@JsonCodec final case class ScalaMainClassesParams(
+final case class ScalaMainClassesParams(
     targets: List[BuildTargetIdentifier],
-    originId: Option[String],
+    originId: Option[String]
 )
 
-@JsonCodec final case class ScalaMainClass(
+object ScalaMainClassesParams {
+  implicit val codec: JsonValueCodec[ScalaMainClassesParams] =
+    JsonCodecMaker.make(CodecMakerConfig.withTransientEmpty(false))
+}
+
+final case class ScalaMainClass(
     `class`: String,
     arguments: List[String],
     jvmOptions: List[String],
     environmentVariables: List[String]
 )
 
-@JsonCodec final case class ScalaMainClassesItem(
+object ScalaMainClass {
+  implicit val codec: JsonValueCodec[ScalaMainClass] =
+    JsonCodecMaker.make(CodecMakerConfig.withTransientEmpty(false))
+}
+
+final case class ScalaMainClassesItem(
     target: BuildTargetIdentifier,
     // Fully qualified names of test classes
     classes: List[ScalaMainClass]
 )
 
-@JsonCodec final case class ScalaMainClassesResult(
+object ScalaMainClassesItem {
+  implicit val codec: JsonValueCodec[ScalaMainClassesItem] =
+    JsonCodecMaker.make(CodecMakerConfig.withTransientEmpty(false))
+}
+
+final case class ScalaMainClassesResult(
     items: List[ScalaMainClassesItem]
 )
 
-@JsonCodec final case class SbtBuildTarget(
+object ScalaMainClassesResult {
+  implicit val codec: JsonValueCodec[ScalaMainClassesResult] =
+    JsonCodecMaker.make(CodecMakerConfig.withTransientEmpty(false))
+}
+
+final case class SbtBuildTarget(
     sbtVersion: String,
     autoImports: List[String],
     scalaBuildTarget: ScalaBuildTarget,
     parent: Option[BuildTargetIdentifier],
-    children: List[BuildTargetIdentifier],
+    children: List[BuildTargetIdentifier]
 )
-@JsonCodec final case class BspConnectionDetails(
+
+object SbtBuildTarget {
+  implicit val codec: JsonValueCodec[SbtBuildTarget] =
+    JsonCodecMaker.make(CodecMakerConfig.withTransientEmpty(false))
+}
+
+final case class BspConnectionDetails(
     name: String,
     argv: List[String],
     version: String,
     bspVersion: String,
     languages: List[String]
 )
 
-@JsonCodec final case class DebugSessionParams(
+object BspConnectionDetails {
+  implicit val codec: JsonValueCodec[BspConnectionDetails] =
+    JsonCodecMaker.make(CodecMakerConfig.withTransientEmpty(false))
```

Comment:
So quick question on this @jvican. Originally you had `withTransientEmpty(true)` for all of these, but I want to verify that it was intentional? When testing, I found that with that, the behavior actually changes from how it was with circe. For example in the `SerializationPropertySuite` you can see that with the old tests and with this set to `false` you'll get something like this where the empty collections are encoded and still there as empty. 
```scala
SerializationPropertySuite.scala:30 bsp4sValueDecoded: BspConnectionDetails("", List(), "", "", List())
SerializationPropertySuite.scala:32 bsp4sJson: "{\"name\":\"\",\"argv\":[],\"version\":\"\",\"bspVersion\":\"\",\"languages\":[]}"
```

However, with `withTransientEmpty` set to `true`, the behavior changes where the empty collections are left out.
```scala
SerializationPropertySuite.scala:30 bsp4sValueDecoded: BspConnectionDetails("", List(), "", "", List())
SerializationPropertySuite.scala:32 bsp4sJson: "{\"name\":\"\",\"version\":\"\",\"bspVersion\":\"\"}"
```

Was this an intentional change that we want to change, or can I keep these as `false` to retain the behavior from before?

Reply:
Ahh ok cool, I've gone ahead and switched these all to use `makeWithRequiredCollectionFields` then 👍🏼